### ISSUE TITLE
Invalid Zip archive: Could not find central directory end

### ISSUE THREAD

**Issue by WeaponMan** [casperstorm/ajour#53]
I have Questie v6.0.4 installed and I wanted to update it to the last version v6.0.5, but its ajour is downloading some botched zip that cannot be open, when I do it from the curseforge website manualy opening the zip works.

Error is "Invalid Zip archive: Could not find central directory end".

Every other addon I had updated succesfully, so maybe something wrong with the curseforge api?

**Comment by casperstorm** [comment]
Thanks for reporting this, @WeaponMan.

I just tested and it worked here on macOS. However, I'll do some more investigation because I did see this error before. The curseforge api is not that stable, but if the https module says the download is succesful then and only then do i send it to be unzipped.

So a few questions, if you have the time:
* Can you upload the botched zip so i can investigate it?
* Can you reproduce it everytime?
* Have you seen this happen on other addons?

<img width="804" alt="Screenshot 2020-08-27 at 12 50 33" src="https://user-images.githubusercontent.com/2248455/91433732-07980b80-e864-11ea-9f17-762b3e12aafc.png">

**Comment by WeaponMan** [comment]
- No need, zip from curseforge have 9.1MB, the downloaded "zip" by ajour doesnt get downloaded all the way randomly ending the stream early
- happens everytime
- I think this problem is only with big addons like questie, weak auras 2 (~11MB) does the same thing if I downgrade it manualy

**Comment by casperstorm** [comment]
Hm, I think this could be the problem: https://github.com/casperstorm/ajour/blob/master/src/network.rs#L18.
I timeout the request if it last longer than 5 seconds, which... seems like a bug. I must have done a typo. Do you mind testing with a larger timeout to see if it fixed the issue?

**Comment by WeaponMan** [comment]
It works now, I thought the timeout is not that "hard" so I didnt try to change that, atleast you could replicate with bigger addons.

**Comment by Sh0ku** [comment]
I've got the same error with updating RaiderIO (Invalid Zip archive: Could not find central directory end).

My Logfile gives this messages:

> 10:44:02.978 [ajour::network][DEBUG] downloading remote version v202010010600 for RaiderIO
> 10:44:18.238 [ajour::gui::update][DEBUG] Message::DownloadedAddon((RaiderIO, error: false))
> 10:44:18.405 [ajour::gui::update][DEBUG] Message::UnpackedAddon((RaiderIO, error: true))

The Zip file itself is not corrupted but its not completely downloaded when this error occures. Sometime it workes fine and the download completes.

Eventually it could help when the download tries to resumeor restart after this error.

**Comment by casperstorm** [comment]
Reopened this issue @Sh0ku.
I'll try to figure out what is going. The one thing that sets RaiderIO aside is the size, its really big compared to other addons.
@tarkah you have any idea what could go wrong? Race condition?

**Comment by Sh0ku** [comment]
Oh sorry i didnt mentioned it, im using Windows 10 2004

**Comment by camas** [comment]
Not sure if this fixes it but you could use `async_std::io::copy` instead of copying the stream yourself. Something like

```rust
async_std::io::copy(body, &mut file).await.expect("TODO: error handling");
```

instead of

https://github.com/casperstorm/ajour/blob/d36f5a6c248dfa9962fae2271694ad9624551336/crates/core/src/network.rs#L84-L99

**Comment by tarkah** [comment]
So I just tested rate limiting my connection to 200kb/s...  it took forever but the download eventually installed and unpacked successfully. So it's definitely not a timeout issue. Something else must be happening to kill the request before the download fully finishes. Not sure why then our request doesn't fail.

**Comment by tarkah** [comment]
I think @WeaponMan 's suggestion of validating our response bytes matches `Content-Length` header makes most sense, and then error out if it doesn't

**Comment by tarkah** [comment]
Ok, I implemented the check on content length in #221 . I also implemented @camas suggestion as that cleaned up the io a bunch.

We still need a way to show install errors in the catalog view, which also relates to #220. So that the user knows the download failed, and can either "retry", or in the instance where no stable file is available for the addon, it shows as "unavailable" or something.
